core/src/Kokkos_OpenACCSpace.hpp

```diff
@@ -0,0 +1,266 @@
+                       const Kokkos::Tools::SpaceHandle =
+                           Kokkos::Tools::make_space_handle(name())) const;
+  friend class Kokkos::Impl::SharedAllocationRecord<
+      Kokkos::Experimental::OpenACCSpace, void>;
+};
+}  // namespace Experimental
+}  // namespace Kokkos
+
+namespace Kokkos {
+namespace Impl {
+
+template <>
+class SharedAllocationRecord<Kokkos::Experimental::OpenACCSpace, void>
+    : public HostInaccessibleSharedAllocationRecordCommon<
+          Kokkos::Experimental::OpenACCSpace> {
+ private:
+  friend class HostInaccessibleSharedAllocationRecordCommon<
+      Kokkos::Experimental::OpenACCSpace>;
+  friend class SharedAllocationRecordCommon<Kokkos::Experimental::OpenACCSpace>;
+  friend Kokkos::Experimental::OpenACCSpace;
+
+  using base_t = HostInaccessibleSharedAllocationRecordCommon<
+      Kokkos::Experimental::OpenACCSpace>;
+  using RecordBase = SharedAllocationRecord<void, void>;
+
+  SharedAllocationRecord(const SharedAllocationRecord&) = delete;
+  SharedAllocationRecord& operator=(const SharedAllocationRecord&) = delete;
+
+  /**\brief  Root record for tracked allocations from this OpenACCSpace
+   * instance */
+  static RecordBase s_root_record;
+
+  const Kokkos::Experimental::OpenACCSpace m_space;
+
+ protected:
+  ~SharedAllocationRecord();
+  SharedAllocationRecord() = default;
+
+  template <typename ExecutionSpace>
+  SharedAllocationRecord(
+      const ExecutionSpace& /*exec_space*/,
+      const Kokkos::Experimental::OpenACCSpace& arg_space,
+      const std::string& arg_label, const size_t arg_alloc_size,
+      const RecordBase::function_type arg_dealloc = &deallocate)
+      : SharedAllocationRecord(arg_space, arg_label, arg_alloc_size,
+                               arg_dealloc) {}
+
+  SharedAllocationRecord(
+      const Kokkos::Experimental::OpenACCSpace& arg_space,
+      const std::string& arg_label, const size_t arg_alloc_size,
+      const RecordBase::function_type arg_dealloc = &deallocate);
+
+ public:
+  KOKKOS_INLINE_FUNCTION static SharedAllocationRecord* allocate(
+      const Kokkos::Experimental::OpenACCSpace& arg_space,
+      const std::string& arg_label, const size_t arg_alloc_size) {
+    if (acc_on_device(acc_device_host)) {
+      return new SharedAllocationRecord(arg_space, arg_label, arg_alloc_size);
+    } else {
+      return nullptr;
+    }
+  }
+};
+
+}  // namespace Impl
+}  // namespace Kokkos
+
+//----------------------------------------------------------------------------
+//----------------------------------------------------------------------------
+
+namespace Kokkos {
+namespace Impl {
+
+// FIXME_OPENACC: implement all possible deep_copies
+template <class ExecutionSpace>
+struct DeepCopy<Kokkos::Experimental::OpenACCSpace,
+                Kokkos::Experimental::OpenACCSpace, ExecutionSpace> {
+  DeepCopy(void* dst, const void* src, size_t n) {
+    if (n > 0) acc_memcpy_device(dst, const_cast<void*>(src), n);
```

Comment:
Do you need to check that `n` is greater than zero?

Reply:
The behavior of `acc_memcpy_device` when `bytes` argument is zero is clarified only in the latest OpenACC specification (V3.2), and thus the value checking is added as a safeguard.  (The current NVHPC (V22.5) supports OpenACC V2.7.)

Reply:
Comments explaining the current DeepCopy implementation are added.